[OpenXR] Call didStartXRSession after xrBeginSession

fujii opened the pull request:
#### 95f37af21285c82c1c728bd7f718db312ad57943
<pre>
[OpenXR] Call didStartXRSession after xrBeginSession
<a href="https://bugs.webkit.org/show_bug.cgi?id=299841">https://bugs.webkit.org/show_bug.cgi?id=299841</a>

Reviewed by Sergio Villar Senin.

&lt;<a href="https://commits.webkit.org/300602@main">https://commits.webkit.org/300602@main</a>&gt; changed to call
didStartXRSession before xrBeginSession. Restored the original
behaviour.

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: <a href="https://commits.webkit.org/301193@main">https://commits.webkit.org/301193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bc748325c55c527613a1b76771b6f5d06bb757a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43399 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34095 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130458 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75829 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44122 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51993 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94047 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62418 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126637 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35145 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110649 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74656 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34108 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73937 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104866 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29032 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133148 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50635 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38548 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102519 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51010 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102359 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47703 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25956 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47468 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19607 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50489 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56251 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49964 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53310 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51638 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5bc748325c55c527613a1b76771b6f5d06bb757a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51540

fujii commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:
```diff
@@ -74,6 +74,7 @@ class OpenXRCoordinator final : public PlatformXRCoordinator {
         WebCore::PageIdentifier pageIdentifier;
         Box<RenderState> renderState;
         RefPtr<WorkQueue> renderQueue;
+        bool didStart { false };
```
`RenderState` is holding a data for the rendering thread.
`didStart` is used only in the main thread.

svillar replied:
True, forget about it then.

fujii commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -643,6 +642,16 @@ void OpenXRCoordinator::handleSessionStateChange()
+                [&](Idle&) { },
+                [&](Active& active) {
+                    if (RefPtr page = WebProcessProxy::webPage(active.pageIdentifier)) {
+                        active.didStart = true;
```
The purpose of `didStart` is remembering didStartXRSession is called so that we can call didEndXRSession only if didStartXRSession is called.
Even if xrBeginSession failed, we call didStartXRSession, so we have to call didEndXRSession.

svillar replied:
Right, my point was more, should we really call didStartXRSession even if xrBeginSession fails. I guess not...

svillar reviewed (approved):
OK, let's go with this. I am still not 100% sure about what we should do with those signals but at least they are coherent.

fujii committed:
[OpenXR] Call didStartXRSession after xrBeginSession
https://bugs.webkit.org/show_bug.cgi?id=299841

Reviewed by Sergio Villar Senin.

<https://commits.webkit.org/300602@main> changed to call
didStartXRSession before xrBeginSession. Restored the original
behaviour.

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: https://commits.webkit.org/301193@main

webkit-commit-queue commented:
Committed 301193@main (95f37af21285): <https://commits.webkit.org/301193@main>

Reviewed commits have been landed. Closing PR #51540 and removing active labels.
